rust/src/mat.rs

```diff
@@ -1,93 +1,201 @@
 use std::ptr;
 
-use ::{Box3, Result, Vec3, VoxelType};
+use {Box3, Result, Vec3, VoxelType};
 
 #[derive(Debug)]
 pub struct Mat<'a> {
     data: &'a mut [u8],
     pub shape: Vec3,
     pub voxel_size: usize,
-    pub voxel_type: VoxelType
+    pub voxel_type: VoxelType,
+    is_fortran_order: bool,
 }
 
 impl<'a> Mat<'a> {
     pub fn new(
         data: &mut [u8],
         shape: Vec3,
         voxel_size: usize,
-        voxel_type: VoxelType
+        voxel_type: VoxelType,
+        is_fortran_array: bool,
     ) -> Result<Mat> {
         // make sure that slice is large enough
         let numel = shape.x as usize * shape.y as usize * shape.z as usize;
         let expected_len = numel * voxel_size;
-
         if data.len() != expected_len {
-            return Err("Length of slice does not match expected size")
+            return Err("Length of slice does not match expected size");
         }
 
         if voxel_size % voxel_type.size() != 0 {
-            return Err("Voxel size must be a multiple of voxel type size")
+            return Err("Voxel size must be a multiple of voxel type size");
         }
 
         Ok(Mat {
             data: data,
             shape: shape,
             voxel_size: voxel_size,
-            voxel_type: voxel_type
+            voxel_type: voxel_type,
+            is_fortran_order: is_fortran_array,
         })
     }
 
-    pub fn as_slice(&self) -> &[u8] { self.data }
-    pub fn as_mut_slice(&mut self) -> &mut [u8] { self.data }
-    pub fn as_mut_ptr(&mut self) -> *mut u8 { self.data.as_mut_ptr() }
+    pub fn get_is_fortran_order(&self) -> bool {
+        self.is_fortran_order
+    }
+    pub fn as_slice(&self) -> &[u8] {
+        self.data
+    }
+    pub fn as_mut_slice(&mut self) -> &mut [u8] {
+        self.data
+    }
+    pub fn as_mut_ptr(&mut self) -> *mut u8 {
+        self.data.as_mut_ptr()
+    }
 
     fn offset(&self, pos: Vec3) -> usize {
-        let offset_vx =
-            pos.x as usize + self.shape.x as usize * (
-            pos.y as usize + self.shape.y as usize * pos.z as usize);
+        let offset_vx = if self.is_fortran_order {
+            pos.x + self.shape.x * (pos.y + self.shape.y * pos.z)
+        } else {
+            pos.z + self.shape.z * (pos.y + self.shape.y * pos.x)
+        };
+        offset_vx as usize * self.voxel_size
+    }
 
-        offset_vx * self.voxel_size
+    pub fn copy_as_fortran_order(&self, buffer: &mut Mat) -> Result<()> {
```

Comment:
Hmm. Would it be possible to make this function "private" and simply call it in `copy_from` when appropriate? Seems confusing to have multiple copy methods.

Reply:
When I move the Fortran conversion into the copy_from, I might have to allocate the `raw_disk_block_buf` for every copy_from call.